File: android/app/src/main/java/com/bunty/clipsync/UpdateChecker.kt
package com.bunty.clipsync

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object UpdateChecker {
    // --- API Configuration ---
    // Fetches latest release tag from GitHub public API
    private const val TAG = "UpdateChecker"
    private const val REPO_OWNER = "WinShell-Bhanu"
    private const val REPO_NAME = "Clipsync"
    private const val GITHUB_API_URL = "https://api.github.com/repos/$REPO_OWNER/$REPO_NAME/releases/latest"

    data class UpdateInfo(
        val version: String,
        val downloadUrl: String,
        val releaseNotes: String
    )

    suspend fun checkForUpdates(currentVersion: String): UpdateInfo? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL(GITHUB_API_URL)
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "GET"
                connection.connectTimeout = 5000
                connection.readTimeout = 5000
                
                // GitHub requires a User-Agent header
                connection.setRequestProperty("User-Agent", "ClipSync-Android-App")

                if (connection.responseCode == 200) {
                    val response = connection.inputStream.bufferedReader().use { it.readText() }
                    val json = JSONObject(response)
                    
                    val latestTag = json.getString("tag_name") // e.g., "v1.0"
                    val htmlUrl = json.getString("html_url")
                    val body = json.optString("body", "New update available!")

                    // Remove 'v' prefix for comparison if present
                    val cleanLatest = latestTag.removePrefix("v")
                    val cleanCurrent = currentVersion.removePrefix("v")

                    if (isVersionNewer(cleanCurrent, cleanLatest)) {
                        Log.d(TAG, "Update found: $latestTag (Current: $currentVersion)")
                        return@withContext UpdateInfo(latestTag, htmlUrl, body)
                    } else {
                        Log.d(TAG, "App is up to date ($currentVersion vs $latestTag)")
                    }
                } else {
                    Log.e(TAG, "GitHub API returned code: ${connection.responseCode}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to check for updates", e)
            }
            return@withContext null
        }
    }

    // --- Version Logic ---
    // Compares semantic versioning strings (e.g. 1.0.0 vs 1.0.1)
    private fun isVersionNewer(current: String, latest: String): Boolean {
        try {
            val currentParts = current.split(".").map { it.toIntOrNull() ?: 0 }
            val latestParts = latest.split(".").map { it.toIntOrNull() ?: 0 }
            
            val length = maxOf(currentParts.size, latestParts.size)
            
            for (i in 0 until length) {
                val c = if (i < currentParts.size) currentParts[i] else 0
                val l = if (i < latestParts.size) latestParts[i] else 0
                
                if (l > c) return true
                if (l < c) return false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Version parsing failed", e)
        }
        return false
    }
}

File: mac/ClipSync/UpdateChecker.swift
import Combine
import Foundation

struct GitHubRelease: Decodable {
    let tag_name: String
    let html_url: String
    let body: String?
}

class UpdateChecker: ObservableObject {
    static let shared = UpdateChecker()

    @Published var updateAvailable: Bool = false
    @Published var latestVersion: String = ""
    @Published var downloadURL: URL?
    @Published var releaseNotes: String = ""

    // --- GitHub API Config ---
    private let repoOwner = "WinShell-Bhanu"
    private let repoName = "Clipsync"
    private let currentVersion = "1.0.0"

    // --- Update Check Logic ---
    func checkForUpdates() {
        // Changed to /releases (list) instead of /releases/latest to support Pre-releases (Betas)
        let urlString = "https://api.github.com/repos/\(repoOwner)/\(repoName)/releases?per_page=1"
        guard let url = URL(string: urlString) else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("OTPSync-Mac-App", forHTTPHeaderField: "User-Agent")

        URLSession.shared.dataTask(with: request) { [weak self] data, response, error in
            guard let data = data, error == nil else {
                print("Update check failed: \(error?.localizedDescription ?? "Unknown error")")
                return
            }

            do {
                // Decode ARRAY of releases, take the first one
                let releases = try JSONDecoder().decode([GitHubRelease].self, from: data)

                if let latestRelease = releases.first {
                    DispatchQueue.main.async {
                        self?.compareVersions(
                            latestTag: latestRelease.tag_name, release: latestRelease)
                    }
                }
            } catch {
                print("Failed to parse release data: \(error)")
                if let str = String(data: data, encoding: .utf8) {
                    print("Raw Response: \(str)")
                }
            }
        }.resume()
    }

    // --- Version Comparison ---
    private func compareVersions(latestTag: String, release: GitHubRelease) {
        let cleanLatest = latestTag.replacingOccurrences(of: "v", with: "")

        // Get actual app version or use fallback
        let appVersion =
            Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? currentVersion
        let cleanCurrent = appVersion.replacingOccurrences(of: "v", with: "")

        if isVersionNewer(current: cleanCurrent, latest: cleanLatest) {
            print("Update available: \(latestTag)")
            self.latestVersion = latestTag
            self.downloadURL = URL(string: release.html_url)
            self.releaseNotes = release.body ?? "New update available!"
            self.updateAvailable = true
        } else {
            print("App is up to date.")
        }
    }

    private func isVersionNewer(current: String, latest: String) -> Bool {
        let currentParts = current.split(separator: ".").compactMap { Int($0) }
        let latestParts = latest.split(separator: ".").compactMap { Int($0) }

        let length = max(currentParts.count, latestParts.count)

        for i in 0..<length {
            let c = i < currentParts.count ? currentParts[i] : 0
            let l = i < latestParts.count ? latestParts[i] : 0

            if l > c { return true }
            if l < c { return false }
        }
        return false
    }
}
